### PR TITLE
[HOPSWORKS-229] dataset api

### PR DESCRIPTION
**Pull request by ErmiasG**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [ ] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by ErmiasG** [comment]
logicalclocks/hops-testing#528

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+      default:
+        throw new WebApplicationException("Action not valid.", Response.Status.NOT_FOUND);
+    }
+    return Response.ok().entity(dto).build();
+  }
+  
+  @POST
+  @Path("{path: .+}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Post an action on a file.")
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java]: @ErmiasG `on a file or dataset`

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    @QueryParam("templateId") Integer templateId, @QueryParam("description") String description,
+    @QueryParam("searchable") Boolean searchable, @QueryParam("generate_readme") Boolean generateReadme,
+    @QueryParam("destination_path") String destPath, @QueryParam("destination_type") DatasetType destDatasetType,
+    @Context UriInfo uriInfo, @Context SecurityContext sc) throws DatasetException, ProjectException,
+    HopsSecurityException {
+    Users user = jwtHelper.getUserPrincipal(sc);
+    DatasetPath datasetPath;
+    DatasetPath distDatasetPath;
+    Project project = this.getProject();
+    RESTApiJsonResponse json = new RESTApiJsonResponse();
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java]: @ErmiasG do we need to use it? I think it's redundant since the status code indicates the result of the operation. Shouldn't it be client creating these messages?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetSortBy.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  private final AbstractFacade.OrderBy param;
+  
+  public DatasetSortBy(String param) {
+    String[] sortByParams = param.split(":");
+    String sort = "";
+    try {
+      sort = sortByParams[0].toUpperCase();
+      this.sortBy = DatasetFacade.Sorts.valueOf(sort);
+    } catch (IllegalArgumentException iae) {
+      throw new WebApplicationException("Sort by need to set a valid sort parameter, but found: " + sort,
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetSortBy.java]: @ErmiasG `need` -> `needs`

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/project/ProjectService.java [hunk 4]
```diff
@@ -707,15 +710,15 @@ public Response quotasByProjectID(
     }
 
     Dataset newDS = new Dataset(inode, destProj);
-    newDS.setShared(true);
+//    newDS.setShared(true);
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/project/ProjectService.java]: @ErmiasG remove comments before merging

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBeanParam.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+public class DatasetBeanParam {
+  
+  @QueryParam("sort_by")
+  @ApiParam(value = "ex. sort_by=ID:asc,date_created:desc",
+    allowableValues = "id:asc, id:desc")
+  private String sortBy;
+  private final Set<DatasetSortBy> sortBySet;
+  private final Set<DatasetSharedWithSortBy> sharedWithSortBySet;
+  @QueryParam("filter_by")
+  @ApiParam(value = "ex. filter_by=id:1, modification_time:date (allowed date formats yyyy-MM-dd'T'HH:mm:ss.SSSZ or " +
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBeanParam.java]: the syntax is add more filter_by
Screenshot from 2019-10-25 10-45-56

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+    if (resourceRequest != null && resourceRequest.contains(ResourceRequest.Name.DATASETS)) {
+      dto.setExpand(true);
+    }
+    return dto;
+  }
+  
+  private DatasetDTO build(DatasetDTO dto, ResourceRequest resourceRequest, Dataset dataset, String parentPath,
+    Users dirOwner, Project project) {
+    expand(dto, resourceRequest);
+    if (dto.isExpand()) {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: expand is allays set if it is called from datasets, the only time the url is returned is when datasets are expandable from another resource.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+      DatasetDTO ownedDatasets = ownedItems(new DatasetDTO(), uriInfo, resourceRequest, project, parentPath, dirOwner);
+      DatasetDTO sharedDatasets =
+        sharedItems(new DatasetDTO(), uriInfo, sharedDatasetResourceRequest, project, parentPath, dirOwner);
+      return mergeAndApplyOffsetAndLimit(dto, resourceRequest, ownedDatasets, sharedDatasets);
+    }
+    return dto;
+  }
+  
+  private DatasetDTO ownedItems(DatasetDTO dto, UriInfo uriInfo, ResourceRequest resourceRequest, Project project,
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: I think the name is self explanatory, for a private method anyway

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java (hunk 1 above)
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java]: file, dir, or dataset

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: Ok then will call ownItems

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/dataset/DatasetSharedWithFacade.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    return em.createNamedQuery("DatasetSharedWith.findByDataset", DatasetSharedWith.class)
+      .setParameter("dataset", dataset).getResultList();
+  }
+  
+  public List<DatasetSharedWith> findByProject(Project project) {
+    return em.createNamedQuery("DatasetSharedWith.findByProject", DatasetSharedWith.class)
+      .setParameter("project", project).getResultList();
+  }
+  
+  public DatasetSharedWith findByProjectAndDataset(Project project, Dataset dataset) {
```
Comment (kai-chi) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/dataset/DatasetSharedWithFacade.java]: this could return an `Optional<DatasetSharedWith>` and be null-safe

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java [hunk 9]
```diff
@@ -133,44 +139,86 @@
...
+    if (project == null) {
+      throw new ProjectException(RESTCodes.ProjectErrorCode.PROJECT_NOT_FOUND, Level.FINE, "projectName: " +
+        projectName);
+    }
+    return project;
+  }
+  
+  private Project getProject() throws ProjectException {
+    return this.projectId != null ? getProjectById() : getProjectByName();
+  }
```
Comment (kai-chi) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java]: see other comments - can you use instead the methods from `projectController`?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java [hunk 10]
```diff
@@ -292,23 +346,19 @@ public Response testMethod(@Context HttpServletRequest request, @Context Securit
...
+    if (info.isUploaded(new ResumableInfo.ResumableChunkNumber(resumableChunkNumber))) {
       json.setSuccessMessage("Uploaded");//This Chunk has been Uploaded.
-      return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK).
-          entity(json).build();
+      return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK).entity(json).build();
     } else {
       json.setErrorMsg("Not uploaded");
-      return noCacheResponse.getNoCacheResponseBuilder(
-          Response.Status.NO_CONTENT).entity(json).build();
+      return noCacheResponse.getNoCacheResponseBuilder(Response.Status.NO_CONTENT).entity(json).build();
```
Comment (kai-chi) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java]: see comments above - I think we are getting rid of `noCacheResponse`. Also refers to other methods in this class

File: hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java [hunk 11]
```diff
@@ -83,21 +83,24 @@
...
 
-  public Dataset uploadToHops(Dataset dataset, String publicDSId) {
-    dataset.setPublicDsState(Dataset.SharedState.HOPS);
+  public Dataset uploadToHops(Project project, Dataset dataset, String publicDSId) {
+    dataset.setPublicDsState(SharedState.HOPS);
     dataset.setPublicDsId(publicDSId);
-    dataset.setEditable(DatasetPermissions.OWNER_ONLY);
+    //TODO:Alex move to set hdfs permissions
+    //datasetController.setPermissions();
+    //dataset.setEditable(DatasetPermissions.OWNER_ONLY);
```
Comment (kai-chi) [inline hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java]: friendly TODO reminder :)

File: hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java [hunk 12]
```diff
@@ -83,21 +83,24 @@
...
   }
   
-  public Dataset unshareFromHops(Dataset dataset) {
-    dataset.setPublicDsState(Dataset.SharedState.PRIVATE);
+  public Dataset unshareFromHops(Project project, Dataset dataset) {
+    dataset.setPublicDsState(SharedState.PRIVATE);
     dataset.setPublicDsId(null);
-    dataset.setEditable(DatasetPermissions.GROUP_WRITABLE_SB);
+    //datasetController.setPermissions();
+    //dataset.setEditable(DatasetPermissions.GROUP_WRITABLE_SB);
```
Comment (kai-chi) [inline hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java]: commented out code can be removed

File: hopsworks-persistence/src/main/resources/META-INF/persistence.xml [hunk 13]
```diff
@@ -151,7 +152,9 @@
     <validation-mode>NONE</validation-mode>
     <properties>
       <property name="eclipselink.target-database" value="MYSQL"/>
-      <!--      <property name="eclipselink.logging.level" value="FINE"/>
+      <!--
+      <property name="eclipselink.profiler" value="PerformanceMonitor"/>
+      <property name="eclipselink.logging.level" value="FINE"/>
```
Comment (kai-chi) [inline hopsworks-persistence/src/main/resources/META-INF/persistence.xml]: what's the purpose of this change?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/hdfs/inode/InodeFacade.java [hunk 14]
```diff
@@ -85,107 +93,35 @@ public InodeFacade() {
...
+    TypedQuery<Inode> query = em.createNamedQuery("Inode.findByParentId", Inode.class);
     query.setParameter("parentId", parent.getId());
     return query.getResultList();
   }
+  
+  public Long countByParentId(Inode parent) {
+    TypedQuery<Long> query = em.createNamedQuery("Inode.countByParentId", Long.class);
+    query.setParameter("parentId", parent.getId());
+    return query.getSingleResult();
+  }
```
Comment (kai-chi) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/hdfs/inode/InodeFacade.java]: @ErmiasG, can you tell me how this method works? I can't find `countByParentId` query on `Inode.class`. How come you invoke it on the `Long.class`?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/hdfs/inode/InodeFacade.java [hunk 15]
```diff
@@ -215,55 +151,15 @@ public Inode findById(Long id) {
       return null;
```
Comment (kai-chi) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/hdfs/inode/InodeFacade.java]: this could also benefit from using `Optional<Inode>`

File: hopsworks-IT/src/test/java/io/hops/hopsworks/dataset/DatasetIT.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+//    try {
+//      assertEquals(templateName, row.getText());
+//    } catch (Error e) {
+//      verificationErrors.append("Should show the new template.").append(e.getMessage());
+//    }
+//    DatasetHelper.gotoDatasetPage(driver);
+//    Dataset dataset = DatasetHelper.createNewDataset(driver, dbHelper, project);
+//    By newDatasetID = By.id("table-row-" + dataset);
+//    Actions action = new Actions(driver);
+//    action.contextClick(driver.findElement(newDatasetID)).perform();
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-IT/src/test/java/io/hops/hopsworks/dataset/DatasetIT.java]: this is not done yet, I commented it out for the test

File: hopsworks-IT/src/test/java/io/hops/hopsworks/dataset/DatasetIT.java [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+  }
+  
+  @Test
+  public void detachMetadata() {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-IT/src/test/java/io/hops/hopsworks/dataset/DatasetIT.java]: Under development not done with this PR

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dataset/DatasetController.java [hunk 18]
```diff
@@ -643,4 +723,371 @@ public void checkFileExists(Path filePath, String username) throws DatasetExcept
...
+      .addCommand(stagingDir)
+      .addCommand(path.toString())
+      .addCommand(hdfsUser)
+      .ignoreOutErrStreams(true)
+      .build();
+    
+    try {
+      ProcessResult processResult = osProcessExecutor.execute(processDescriptor);
+      int result = processResult.getExitCode();
+      if (result == 2) {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dataset/DatasetController.java]: This code is just moved from the rest api, and I think it is not a common exit code

File: hopsworks-IT/src/test/java/io/hops/hopsworks/util/models/Dataset.java [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+  private Integer id;
+  private Long inodePid;
+  private String inodeName;
+  private Long partitionId;
+  private String name;
+  private String description;
+  private boolean searchable;
+  private int publicDs;
+  private String publicDsId;
+  private String dstype;
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-IT/src/test/java/io/hops/hopsworks/util/models/Dataset.java]: auto generated name, but I can change it.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+    @QueryParam("action") DatasetActions.Get action, @QueryParam("mode") FilePreviewMode mode,
+    @BeanParam Pagination pagination, @BeanParam InodeBeanParam inodeBeanParam,
+    @BeanParam DatasetExpansionBeanParam datasetExpansionBeanParam,
+    @Context UriInfo uriInfo, @Context SecurityContext sc) throws DatasetException, ProjectException {
+    Users user = jwtHelper.getUserPrincipal(sc);
+    ResourceRequest resourceRequest = new ResourceRequest(ResourceRequest.Name.INODES);
+    Project project = this.getProject();
+    DatasetPath datasetPath = datasetHelper.getDatasetPathIfFileExist(project, path, datasetType);
+    InodeDTO dto;
+    switch (action == null? DatasetActions.Get.STAT : action) {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java]: The business logic should be pushed to the controller, but this is api specific work

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+        resourceRequest.setFilter(inodeBeanParam.getFilter());
+        dto = inodeBuilder.buildItems(uriInfo, resourceRequest, project, datasetPath);
+        break;
+      case STAT:
+        if (datasetPath.isTopLevelDataset()) {
+          ResourceRequest datasetResourceRequest = new ResourceRequest(ResourceRequest.Name.DATASETS);
+          datasetResourceRequest.setExpansions(datasetExpansionBeanParam.getResources());
+          DatasetDTO datasetDTO = datasetBuilder.build(uriInfo, datasetResourceRequest, project,
+            datasetPath.getDataset(), null, null);
+          return Response.ok().entity(datasetDTO).build();
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetResource.java]: It is not recommended but I think it is better than having two different end-points just for the return type.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/project/util/PathValidator.java [hunk 22]
```diff
@@ -188,25 +194,20 @@ private void buildProjectDsRelativePath(Project project, String[] pathComponents
   private void buildHiveDsRelativePath(Project project, String[] pathComponents, DsPath dsPath)
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/project/util/PathValidator.java]: This class will be removed soon

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/util/DownloadService.java [hunk 23]
```diff
@@ -86,41 +96,105 @@
...
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Get a one time download token.", response = RESTApiJsonResponse.class)
+  @AllowedProjectRoles({AllowedProjectRoles.DATA_SCIENTIST, AllowedProjectRoles.DATA_OWNER})
+  @JWTRequired(acceptedTokens={Audience.API}, allowedUserRoles={"HOPS_ADMIN", "HOPS_USER"})
+  public Response getDownloadToken(@PathParam("path") String path, @QueryParam("type") DatasetType datasetType,
+    @Context SecurityContext sc) throws DatasetException, ProjectException {
+    if(!settings.isDownloadAllowed()){
+      throw new DatasetException(RESTCodes.DatasetErrorCode.DOWNLOAD_NOT_ALLOWED, Level.FINEST);
+    }
+    Users user = jWTHelper.getUserPrincipal(sc);
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/util/DownloadService.java]: do not see it being used anywhere else,

File: hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java (hunk 12 above)
Comment (ErmiasG) [reply hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java]: Alex wanted a reminder to fix it later

File: hopsworks-persistence/src/main/resources/META-INF/persistence.xml (hunk 13 above)
Comment (ErmiasG) [reply hopsworks-persistence/src/main/resources/META-INF/persistence.xml]: To enable it when we want to monitor sql performance

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dataset/DatasetController.java [hunk 24]
```diff
@@ -643,4 +723,371 @@ public void checkFileExists(Path filePath, String username) throws DatasetExcept
...
+      datasetRequest.remove(dsReq);//the dataset is shared so remove the request.
+    }
+    
+    activityFacade
+      .persistActivity(ActivityFacade.SHARED_DATA + ds.getName() + " with project " + targetProject.getName(),
+        project, user, ActivityFlag.DATASET);
+  }
+  
+  public void acceptShared(Project project, DatasetSharedWith datasetSharedWith) throws DatasetException {
+    if (datasetSharedWith == null) {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dataset/DatasetController.java]: Can be done on another PR, we can start a discussion on optional

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java [hunk 25]
```diff
@@ -0,0 +1,409 @@
...
+      for (; sort.hasNext() && c == 0; ) {
+        c = compare(a, b, sort.next());
+      }
+      return c;
+    }
+  }
+  
+  private DatasetDTO mergeAndApplyOffsetAndLimit(DatasetDTO dto, ResourceRequest resourceRequest, DatasetDTO dto1,
+    DatasetDTO dto2) {
+    if ((dto1 == null || dto1.getItems() == null || dto1.getItems().isEmpty()) &&
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: Count can be > 0 even when the list is empty if limit is set

[the remark this answers is not in the document]
Comment (kai-chi) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: so `count` is not the number of items in the `items` field?
actually, you are setting `count` to the number of items after limit so, in dataset case, it will always apply (https://github.com/ErmiasG/hopsworks/blob/458dc77eff7858d98f6282a4cf935c4d8a453d1d/hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java#L407) but I gt your point. maybe it doesn't apply for other cases

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/project/MetadataService.java [hunk 26]
```diff
@@ -439,18 +434,43 @@ public Response detachTemplateFromInode(@PathParam("inodeid") Long inodeid,
...
+    Inode inode = inodeController.getInodeAtPath(inodePath);
+    Template temp = templatefacade.findByTemplateId(templateid);
+    temp.getInodes().add(inode);
+    
+    //persist the relationship
+    this.templatefacade.updateTemplatesInodesMxN(temp);
+    
+    RESTApiJsonResponse json = new RESTApiJsonResponse();
+    json.setSuccessMessage("The template was attached to file " + inode.getId());
+    return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK).entity(json).build();
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/project/MetadataService.java]: This code is moved from dataset service, we can remove it when we do the metadata service

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java (hunk 10 above)
Comment (ErmiasG) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/util/UploadService.java]: Not using it on new end-points, this one is not new

File: hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java (hunk 11 above)
Comment (ErmiasG) [reply hopsworks-dela/src/main/java/io/hops/hopsworks/dela/DelaDatasetController.java]: Alex wanted a reminder to fix it later,

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/dataset/DatasetBuilder.java]: `count` is the number of all instances of this resource after the request filters are applied. For example, if the request asks to filter all datasets with name like `abc`, and that returns 10 items then `count` will be 10. But if on the same request we say offset=5 limit=5, then 5 items will be returned but count will be 10. @ErmiasG correct me if I'm wrong :)
